test-app/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt

```diff
@@ -0,0 +1,437 @@
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun initNavigation() {
+        mapboxNavigation = MapboxNavigation(
+            NavigationOptions.Builder(this)
+                .accessToken(getMapboxAccessTokenFromResources())
+                .locationEngine(LocationEngineProvider.getBestLocationEngine(this))
+                .build()
+        ).apply {
+            registerLocationObserver(object : LocationObserver {
+                override fun onRawLocationChanged(rawLocation: Location) {
+                    updateCameraToIdle()
+                    val point = Point.fromLngLat(rawLocation.longitude, rawLocation.latitude)
+                    val cameraOptions = CameraOptions.Builder()
+                        .center(point)
+                        .zoom(13.0)
+                        .pitch(45.0)
+                        .build()
+                    mapboxMap.easeTo(cameraOptions)
+                    navigationLocationProvider.changePosition(rawLocation)
+                    mapboxNavigation.unregisterLocationObserver(this)
+                }
+
+                override fun onEnhancedLocationChanged(
+                    enhancedLocation: Location,
+                    keyPoints: List<Location>
+                ) {
+                }
+            })
+        }
+        mapboxNavigation.startTripSession()
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun initStyle() {
+        mapboxMap.loadStyleUri(
+            Style.MAPBOX_STREETS,
+            {
+                getGesturePlugin().addOnMapLongClickListener(this)
+            },
+            object : OnMapLoadErrorListener {
+                override fun onMapLoadError(mapViewLoadError: MapLoadError, msg: String) {
+                }
+            }
+        )
+    }
+
+    private fun findRoute(origin: Point, destination: Point) {
+        mapboxNavigation.requestRoutes(
+            RouteOptions.builder()
+                .applyDefaultParams()
+                .accessToken(getMapboxAccessTokenFromResources())
+                .coordinates(listOf(origin, destination))
+                .alternatives(true)
+                .build()
+        )
+    }
+
+    private fun updateCameraToOverview() {
+        viewportDataSource.onRouteChanged(directionsRoute)
+        viewportDataSource.overviewPaddingPropertyOverride(overviewEdgeInsets)
+        viewportDataSource.evaluate()
+        navigationCamera.requestNavigationCameraToOverview()
+    }
+
+    private fun updateCameraToFollowing() {
+        viewportDataSource.followingPaddingPropertyOverride(followingEdgeInsets)
+        viewportDataSource.evaluate()
+        navigationCamera.requestNavigationCameraToFollowing()
+    }
+
+    private fun updateCameraToIdle() {
+        navigationCamera.requestNavigationCameraToIdle()
+    }
+
+    private fun startNavigation() {
+        updateCameraToFollowing()
+        mapboxNavigation.registerLocationObserver(locationObserver)
```

Comment:
We cannot register this observer only after the "start navigation button" is clicked because this introduces a couple of issues:
1. The puck is not being updated until we click the start button.
2. The viewport data source is not being fed the data, so when the button is clicked we "requestNavigationCameraToFollowing" but the viewport system has no data so it transitions to the null island only after the next location updates moves back to the puck.